src/coreclr/gc/gc.cpp

```diff
@@ -20926,16 +20926,29 @@ void gc_heap::enque_pinned_plug (uint8_t* plug,
 
     if (save_pre_plug_info_p)
     {
-#ifdef SHORT_PLUGS
-        BOOL is_padded = is_plug_padded (last_object_in_last_plug);
-        if (is_padded)
-            clear_plug_padded (last_object_in_last_plug);
-#endif //SHORT_PLUGS
+#if defined(SHORT_PLUGS) || defined(DOUBLY_LINKED_FL)
+        // In the case of short plugs or doubly linked free lists, there may be extra bits
+        // set in the method table pointer.
+        // Clear these bits for the copy saved in saved_pre_plug, but not for the copy
+        // saved in saved_pre_plug_reloc.
+        // This is because we need these bits for compaction, but not for mark & sweep.
+        MethodTable* pRawMethodTable = header(last_object_in_last_plug)->RawGetMethodTable();
+        MethodTable* pMethodTable = header(last_object_in_last_plug)->GetMethodTable();
+        if (pRawMethodTable != pMethodTable)
+        {
+            header(last_object_in_last_plug)->RawSetMethodTable (pMethodTable);
+            dprintf (3, ("method table for plug %Ix has low bits set: %Ix",
+                last_object_in_last_plug,
+                ((size_t)pRawMethodTable) - ((size_t)pMethodTable)));
+        }
+#endif //SHORT_PLUGS || DOUBLY_LINKED_FL
+        // now copy the bits over
```

Comment:
it would be a bit nicer to follow the current coding convention, perhaps something like?

```
inline
size_t clear_special_bits_mt (uint8_t* obj)
{
#if defined(SHORT_PLUGS) || defined(DOUBLY_LINKED_FL)
    size_t special_bits = header(node)->ClearSpecialBits();
    // could also assert to make sure no special bits are set if neither define is defined
#endif //SHORT_PLUGS || DOUBLY_LINKED_FL
}

inline
void set_special_bits_mt (uint8_t* obj, size_t special_bits)
{
#if defined(SHORT_PLUGS) || defined(DOUBLY_LINKED_FL)
    header(node)->SetSpecialBits (special_bits);
#endif //SHORT_PLUGS || DOUBLY_LINKED_FL
}

size_t special_bits = clear_special_bits_mt (last_object_in_last_plug);
memcpy (&(m.saved_post_plug), m.saved_post_plug_info_start, sizeof (gap_reloc_pair));
set_special_bits_mt (last_object_in_last_plug, special_bits);
```